R/accounts.R

```diff
@@ -60,74 +60,121 @@ connectApiUser <- function(account = NULL, server = NULL, apiKey, quiet = FALSE)
+                        server = NULL,
+                        quiet = FALSE,
                         launch.browser = getOption("rsconnect.launch.browser", interactive())) {
   server <- findServer(server)
+  resp <- getAuthTokenAndUser(server, launch.browser)
+
+  registerAccount(
+    serverName = server,
+    accountName = resp$user$username,
+    accountId = resp$user$id,
+    token = resp$token$token,
+    private_key = resp$token$private_key
+  )
 
-  # generate a token and send it to the server
-  token <- getAuthToken(server)
   if (!quiet) {
-    message("A browser window should open; if it doesn't, you may authenticate ",
-            "manually by visiting ", token$claim_url, ".")
-    message("Waiting for authentication...")
+    accountId <- accountId(resp$user$username, server)
+    cli::cli_alert_success("Registered account for {accountId}")
   }
+  invisible()
+}
+
+getAuthTokenAndUser <- function(server, launch.browser = TRUE) {
+  # Generate public/private key pair
+  token <- generateToken()
+
+  # Send public key to server, and generate URL where the token can be claimed
+  client <- clientForAccount(list(server = server))
+  response <- client$addToken(list(
+    token = token$token,
+    public_key = token$public_key,
+    user_id = 0L
+  ))
+
+  claim_url <- response$token_claim_url
 
   if (isTRUE(launch.browser))
-    utils::browseURL(token$claim_url)
+    utils::browseURL(claim_url)
   else if (is.function(launch.browser))
-    launch.browser(token$claim_url)
+    launch.browser(claim_url)
 
-  # keep trying to authenticate until we're successful
-  repeat {
-    Sys.sleep(1)
-    user <- getAuthedUser(server, token = token)
-    if (!is.null(user))
-      break
+  if (isFALSE(launch.browser)) {
+    cli::cli_alert_warning("Open {.url {claim_url}} to authenticate")
+  } else {
+    cli::cli_alert_info("A browser window should open to complete authentication")
+    cli::cli_alert_warning("If doesn't open, then visit {.url {claim_url}}")
   }
 
-  # populate the username if there wasn't one set on the server
-  if (nchar(user$username) == 0) {
-    if (!is.null(account))
-      user$username <- account
-    else
-      user$username <- tolower(paste0(substr(user$first_name, 1, 1),
-                                      user$last_name))
-
-    # in interactive mode, prompt for a username before accepting defaults
-    if (!quiet && interactive() && is.null(account)) {
-      input <- readline(paste0("Choose a nickname for this account (default '",
-                               user$username, "'): "))
-      if (nchar(input) > 0)
-        user$username <- input
+  # keep trying to authenticate until we're successful; server returns
+  # 500 "Token is unclaimed error" while waiting for interactive auth to complete
+  cli::cli_progress_bar(format = "{cli::pb_spin} Waiting for authentication...")
+  repeat {
+    for (i in 1:10) {
+      Sys.sleep(0.1)
+      cli::cli_progress_update()
```

Comment:
Why 10 progress updates between attempts? Maybe one progress update per attempt with ~1s between attempts?

Reply:
Because then you get a cool spinner 😄